D-Freezer/freezer.md

```diff
@@ -0,0 +1,42 @@
+# Implementing a scheduling class
+
+## Introduction
+One of the reasons students struggle with this assignment boils down to a lack of understanding of how the scheduler works. In this guide, I hope to provide you with a clear understanding of how the Linux scheduler pieces fit together. I hope to paint a picture that you can use to help you implement the freezer scheduler.
+
+## The `task_struct`
+
+Recall from HW1 that in Linux, every process is defined by its `struct task_struct`. When you have multiple tasks forked off a common parent, they are linked together in a doubly linked-list `struct list_head siblings` embedded within the `task_struct`. For example, if you had four processes running on your system, each forked off one parent, it will look something like this (the parent is not shown): 
+
+<div align='center'>
+    <img src='./task_struct.png'/><br/>
+</div>
+
+However, at this stage, none of these processes are actually running on a CPU. In order to get them onto a CPU, I need to introduce you to the `struct rq`. 
+
+## The `struct_rq`
+
+The `struct rq` is a per-cpu run queue data structure. I like to think of it as the virtual CPU. It contains a lot of information (must of which goes way over my head), but it also includes the list of tasks that will (eventually) run on that CPU. 
+So how does that work? 
+A naive implementation would be to embed a `struct list_head runqueue_head` (for example) into the `struct rq`, and embed a `struct list_head node` into every `task_struct`. 
+<div align='center'>
+    <img src='./naive.png'/><br/>
+    This is BAD implementation. 
+</div
```

Comment:
`</div>`.
The following paragraph doesn't show up in the markdown because the closing tag is missing a `>` lol

Reply:
done